### PR TITLE
Add receiver for sapm protocol

### PR DESCRIPTION
**Pull request by codesmith14**

### CODE REVIEW

**Comment by codesmith14** [comment]
I think the ci is failing because `sapm-proto` is still private.  @tigrannajaryan could you please make `sapm-proto` public?

File: receiver/sapmreceiver/factory.go
Comment (owais) [inline receiver/sapmreceiver/factory.go]: I've never seen this before. _I think_ this is supposed to be used internally by the pipeline and not the factory. If a component is disabled, pipeline should never ask it's factory to create a receiver.

File: receiver/sapmreceiver/trace_receiver.go
Comment (owais) [inline receiver/sapmreceiver/trace_receiver.go]: Otel has a package to combine multiple errors: https://github.com/open-telemetry/opentelemetry-collector/blob/master/oterr/errors.go

Comment (codesmith14) [reply receiver/sapmreceiver/trace_receiver.go]: cool, I copied that from the jaeger receiver.  I'll update to use that.

File: go.sum
Comment (tigrannajaryan) [inline go.sum]: Is it valid to consume 2 different versions of the same lib? Which one is actually used by "import" statements?

File: go.sum
Comment (tigrannajaryan) [inline go.sum]: Same here.

File: go.sum
Comment (tigrannajaryan) [inline go.sum]: I think we need to make this repo public, otherwise this won't build.

File: receiver/sapmreceiver/config.go
Comment (tigrannajaryan) [inline receiver/sapmreceiver/config.go]:
```suggestion
// Config defines configuration for SAPM receiver.
```

File: receiver/sapmreceiver/config.go
Comment (tigrannajaryan) [inline receiver/sapmreceiver/config.go]: I think this is not necessary, these all are already implemented in `configmodels.ReceiverSettings`.

File: receiver/sapmreceiver/trace_receiver.go
Comment (owais) [inline receiver/sapmreceiver/trace_receiver.go]: Shouldn't this be using the per request context? In fact, I think it should use `req.Context()`.

File: receiver/sapmreceiver/factory.go
Comment (tigrannajaryan) [inline receiver/sapmreceiver/factory.go]: We should probably make this a utility func available for all components. No need to change anything in this PR, maybe just a TODO note for future improvement.

File: receiver/sapmreceiver/factory.go
Comment (tigrannajaryan) [reply receiver/sapmreceiver/factory.go]: Yes, this should be handled by https://github.com/open-telemetry/opentelemetry-collector/blob/d32cd2a100a49d2dcd34ba4401d50b4ee15adc37/config/config.go#L711
If it is not then we have a bug.

File: receiver/sapmreceiver/trace_receiver.go
Comment (codesmith14) [reply receiver/sapmreceiver/trace_receiver.go]: This is the main thing I wanted to talk about

File: receiver/sapmreceiver/trace_receiver.go
Comment (tigrannajaryan) [inline receiver/sapmreceiver/trace_receiver.go]: I think `ConsumeTraceData` can return an error, it needs to be handled here otherwise we may have silent drops.

File: receiver/sapmreceiver/trace_receiver.go
Comment (tigrannajaryan) [inline receiver/sapmreceiver/trace_receiver.go]: @owais @pjanotti Do we have own trace collection enabled in the Collector? Do these spans go anywhere or it is a noop currently?

File: receiver/sapmreceiver/trace_receiver.go
Comment (tigrannajaryan) [inline receiver/sapmreceiver/trace_receiver.go]: Shouldn't we `return` here? Is there a point in continuing and trying to setup the http.Server?

**Comment by tigrannajaryan** [comment]
Nit: capitalize subject line of commit message.
(At Omnition we were typically following this advice: https://chris.beams.io/posts/git-commit/)

**Comment by codesmith14** [comment]
After discussing with @owais, I'm going to remove the handler from sapm-proto.  I'll leave the request parsing function, but the handler itself will be moved into the otel receiver directly.

File: receiver/sapmreceiver/trace_receiver.go
Comment (codesmith14) [reply receiver/sapmreceiver/trace_receiver.go]: @tigrannajaryan Is returning on error appropriate for converting the batches and consuming their trace data?  Or should I attempt to process the rest of the batches before returning?

File: receiver/sapmreceiver/factory.go
Comment (codesmith14) [reply receiver/sapmreceiver/factory.go]: removed

File: receiver/sapmreceiver/trace_receiver.go
Comment (codesmith14) [reply receiver/sapmreceiver/trace_receiver.go]: didn't actually have multiple errors any way.  It was a bad copy/paste.  I've removed it entirely.

File: receiver/sapmreceiver/config.go
Comment (codesmith14) [reply receiver/sapmreceiver/config.go]: removed

File: receiver/sapmreceiver/factory.go
Comment (codesmith14) [reply receiver/sapmreceiver/factory.go]: added a TODO

File: receiver/sapmreceiver/trace_receiver.go
Comment (codesmith14) [reply receiver/sapmreceiver/trace_receiver.go]: added a return

File: go.sum
Comment (owais) [reply go.sum]: This is not really consuming. It's just recording the checksum for the version so it it sees the version again, it'll reject it if the checksum does not match. Probably some dependency is using 1.3.1 and some other 1.3.0 ?

File: receiver/sapmreceiver/trace_receiver.go
Comment (tigrannajaryan) [reply receiver/sapmreceiver/trace_receiver.go]: We have 2 options:
1. Return ErrBadRequest error to the caller. This will ensure caller does not re-send this anymore.
2. Return "Success" but increment local "bad input data" counter and process the rest of batches which are good.

(1) is definitely more appropriate if ProtoBatchToOCProto returns an error because it typically means something is wrong with span data.
(2) _may_ be more appropriate for `ConsumeTraceData` errors, but I could argue either way.

To keep it simple I would handle errors from both functions the same way and would just return ErrBadRequest (but make sure we actually return this exact error so that SAPM responds with 400 Bad Data and we don't receive this data again).

File: go.sum
Comment (tigrannajaryan) [reply go.sum]: OK, so this means that according to "Minimal Version Selection" rule 1.3.1 will win.

Comment (owais) [reply go.sum]: Actually 1.3.2 wins. Which exact version of a dependency is used is only dictated by go.mod. go.sum only contains checksums. Relevant line: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/48/files#diff-c391b461a7bdb4bc745d488660866ab0R7

**Comment by codesmith14** [comment]
@tigrannajaryan @owais I think I've addressed all comments on this pr.  I'll update the sapm-proto dependency with a tagged version once this pr is approved: https://github.com/signalfx/sapm-proto/pull/7

**Review by pjanotti — COMMENTED** [review]
Looks good in general. I miss somewhere a description for SAPM (Splunk APM?), best place will be adding a doc.go for the package and describe it there.

File: receiver/sapmreceiver/trace_receiver.go
Comment (codesmith14) [reply receiver/sapmreceiver/trace_receiver.go]: I guess it can't, I'll remove the nil check

File: receiver/sapmreceiver/factory.go
Comment (codesmith14) [reply receiver/sapmreceiver/factory.go]: I copied this check from the jaeger receiver.  I think it's primary purpose is to prevent a 0 valued port from being passed to the listener.  This wouldn't raise an error it would just assign a random port to the listener.

File: receiver/sapmreceiver/trace_receiver.go
Comment (codesmith14) [reply receiver/sapmreceiver/trace_receiver.go]: I think the thinking is that in the future this might return something.  I'll go ahead and make a default response in the receiver's constructor and send that for now.

File: receiver/sapmreceiver/trace_receiver.go
Comment (codesmith14) [reply receiver/sapmreceiver/trace_receiver.go]: The naked return is for the function passed into `sr.startOnce.Do()`, the error that is being set in the line above is declared outside of this inner function.  So ultimately we will return the error and not nil.

File: receiver/sapmreceiver/trace_receiver.go
Comment (owais) [reply receiver/sapmreceiver/trace_receiver.go]: I think we should do it if acceptEndocingHeader is present. At least it documents what the behavior should be. As an optimization, until we actually need a real response, we can have the client never ask for gzipped response. Either that or the encoding code below can just check if the response to be sent is empty and skip encoding.

**Review by tigrannajaryan — APPROVED** [review]
LGTM on my part, please address remaining comments from other reviewers.

File: receiver/sapmreceiver/trace_receiver.go
Comment (codesmith14) [reply receiver/sapmreceiver/trace_receiver.go]: Ok cool, then I'm just going to have a default response in []byte.  And we'll still pass that through the gzip code here.  If we ever decide to put stuff in the response we can revise this function to actually build the response byte array before going through this gzip path.

**Review by owais — APPROVED** [review]

File: receiver/sapmreceiver/trace_receiver_test.go
Comment (codesmith14) [reply receiver/sapmreceiver/trace_receiver_test.go]: I can't use that because that's in a package internal to `optentelemetry-collector` and isn't in `opentelemetry-collector-contrib`.  But since this is the default endpoint, i'll use the constant in this package.

File: receiver/sapmreceiver/trace_receiver.go
Comment (pjanotti) [reply receiver/sapmreceiver/trace_receiver.go]: ops, missed that

File: receiver/sapmreceiver/trace_receiver_test.go
Comment (pjanotti) [reply receiver/sapmreceiver/trace_receiver_test.go]: I will take care of exposing that soon (no need to hold this PR for that).

File: receiver/sapmreceiver/factory.go
Comment (pjanotti) [reply receiver/sapmreceiver/factory.go]: hum, perhaps we need a better and general way to handle that. Anyway, it is a separated issue.

File: receiver/sapmreceiver/trace_receiver.go
Comment (codesmith14) [reply receiver/sapmreceiver/trace_receiver.go]: Will do.  @tigrannajaryan is there some kind of convention in otel for handling this?  I don't think zap has a rate limited logger (at least i couldn't find one in their docs).  I hesitate to put a plain log statement there because log files could blow up from a bad actor.

Comment (codesmith14) [reply receiver/sapmreceiver/trace_receiver.go]: hrm I think i found something in zap. They refer to this as "sampling".  I'll have to read their doc further to understand how to use that feature.  I've never used zap before.

**Comment by codesmith14** [comment]
I'll continue looking into the logging stuff separately and open another pr to address it.  @tigrannajaryan @pjanotti @owais I think this pr is ready to merge pending your final approval.

**Review by pjanotti — APPROVED** [review]
Thanks @codesmith14 LGTM

File: receiver/sapmreceiver/trace_receiver.go
Comment (tigrannajaryan) [reply receiver/sapmreceiver/trace_receiver.go]: I think it is best to record a metric instead of logging. I believe that is how we expose most of bad things happening in Collector that can happen because of input data.

**Pull request merged**
